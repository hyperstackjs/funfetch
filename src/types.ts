export interface ParsedResponse {
  body: string | any
  status: number
  headers: Record<string, string>
}
export type SimpleResponse = string | any

export interface MethodConfig {
  bearer?: string
  basic?: { user: string; password: string }
  base?: string
  fetch?: any
  baseOpts?: RequestInit
  fullResponse?: boolean
  responseHandler?: (res: Response) => ParsedResponse | SimpleResponse
  throwIfBadStatus?: (res: Response) => void
  throwIfError?: (res: ParsedResponse | SimpleResponse) => void
}
export type Endpoint<TParams, TData, TQuery, TRes> = (
  data?: EndpointData<TParams, TData, TQuery>
) => Promise<TRes>

export type DynamicEndpoint = Endpoint<any, any, any, any>

export interface EndpointData<TParams, TData, TQuery> {
  params?: TParams | any
  data?: TData | any
  query?: TQuery | any
  opts?: RequestInit
}
export interface EndpointParams {
  ep: any
  config?: MethodConfig
}

export interface RequestParams<TParams, TData, TQuery> {
  method: string
  params?: TParams
  data?: TData
  query?: TQuery
  baseOpts?: RequestInit
  opts?: RequestInit
}

export interface Funfetch {
  post: <TParams, TData, TQuery, TRes>(
    ep: string | ((params: any) => string)
  ) => Endpoint<TParams, TData, TQuery, TRes>
  get: <TParams, TData, TQuery, TRes>(
    ep: string | ((params: any) => string)
  ) => Endpoint<TParams, TData, TQuery, TRes>
  put: <TParams, TData, TQuery, TRes>(
    ep: string | ((params: any) => string)
  ) => Endpoint<TParams, TData, TQuery, TRes>
  del: <TParams, TData, TQuery, TRes>(
    ep: string | ((params: any) => string)
  ) => Endpoint<TParams, TData, TQuery, TRes>
}
