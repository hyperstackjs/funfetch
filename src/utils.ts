export const headersToObject = (headers: Headers) => {
  const acc: Record<string, string> = {}
  for (const p of headers.entries()) {
    acc[p[0]] = p[1]
  }
  return acc
}
