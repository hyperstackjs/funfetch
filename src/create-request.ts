import isFunction from 'lodash/isFunction'
import { Headers } from 'cross-fetch'
import merge from 'lodash/merge'
import type { EndpointParams, RequestParams } from './types'
import { headersToObject } from './utils'

const buildUrl = <TQuery>(targetUrl: string, query: TQuery, base?: string) => {
  // use a dummy example.com to avoid an exception, and in order
  // to later manipulate params via a fully parsed URL
  const newUrl = new URL(targetUrl, base || 'http://example.com')
  if (query) {
    Object.entries(query).forEach(([k, v]: [string, any]) =>
      newUrl.searchParams.append(k, v)
    )
  }

  // not an absolute URL originally, no one supplied a base,
  // this means we need to strip 'example.com'
  if (!targetUrl.startsWith('http') && !base) {
    return newUrl.pathname + newUrl.search + newUrl.hash
  }
  return newUrl.toString()
}

export const createRequest = <TParams, TData, TQuery>(
  { ep, config }: EndpointParams,
  {
    method,
    params,
    data,
    query,
    baseOpts,
    opts,
  }: RequestParams<TParams, TData, TQuery>
) => {
  const headers = new Headers({
    'Content-Type': 'application/json',
  })
  if (config?.basic) {
    headers.set(
      'Authorization',
      `Basic ${btoa(`${config.basic.user}:${config.basic.password}`)}`
    )
  }
  if (config?.bearer) {
    headers.set('Authorization', `Bearer ${config.bearer}`)
  }

  const customopts: RequestInit = {}
  customopts.headers = headersToObject(headers)

  const producedUrl = isFunction(ep) ? ep(params) : ep
  const targetUrl = buildUrl(producedUrl, query, config?.base)

  return {
    url: targetUrl,
    body: data ? JSON.stringify(data) : undefined,
    method,
    opts: merge({}, baseOpts, customopts, opts),
  }
}
