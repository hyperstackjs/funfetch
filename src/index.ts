import fetch from 'cross-fetch'
import createDebug from 'debug'
import type {
  Endpoint,
  EndpointData,
  Funfetch,
  MethodConfig,
  ParsedResponse,
  SimpleResponse,
} from './types'
import { createRequest } from './create-request'
import { headersToObject } from './utils'

export * from './types'

const debug = createDebug('funfetch')

const noop = (_res: ParsedResponse | SimpleResponse) => {}

const fullResponseJsonHandler = async (
  x: Response
): Promise<ParsedResponse> => {
  const { headers } = x
  const contentType = headers.get('Content-Type')
  if (contentType?.includes('application/json')) {
    return {
      body: await x.json(),
      status: x.status,
      headers: headersToObject(x.headers),
    }
  }
  return {
    body: await x.text(),
    status: x.status,
    headers: headersToObject(x.headers),
  }
}
const simpleResponseJsonHandler = async (
  x: Response
): Promise<SimpleResponse> => {
  const { headers } = x
  const contentType = headers.get('Content-Type')
  if (contentType?.includes('application/json')) {
    return await x.json()
  }
  return await x.text()
}

const throwIfBadStatus = (response: Response) => {
  if (response.status >= 400 && response.status < 600) {
    throw new Error(JSON.stringify(response))
  }
}

export const createMethod =
  (method: string, config?: MethodConfig) =>
  <TParams, TData, TQuery, TRes>(
    ep: string | ((params: TParams | any) => string)
  ): Endpoint<TParams, TData, TQuery, TRes> =>
  async ({
    params,
    data,
    query,
    opts,
  }: EndpointData<TParams, TData, TQuery> = {}): Promise<TRes> => {
    const fetcher = config?.fetch || fetch
    const req = createRequest<TParams, TData, TQuery>(
      { ep, config },
      { method, params, data, query, opts, baseOpts: config?.baseOpts }
    )
    debug('request: %j', req)
    const res = await fetcher(req.url, {
      method: req.method,
      body: req.body,
      ...req.opts,
    })

    const throwIfBad = config?.throwIfBadStatus || throwIfBadStatus
    throwIfBad(res)

    const convertedResponse = await (
      config?.responseHandler ||
      (config?.fullResponse
        ? fullResponseJsonHandler
        : simpleResponseJsonHandler)
    )(res)

    // parsed response can be: ParsedResponse (full), string (text()), any (json())
    debug('response: %O', convertedResponse)

    const throwIfError = config?.throwIfError || noop
    throwIfError(convertedResponse)

    debug('done: (no errors)')
    return convertedResponse
  }

export const funfetch = (config: MethodConfig = {}): Funfetch => {
  return {
    post: createMethod('POST', config),
    get: createMethod('GET', config),
    put: createMethod('PUT', config),
    del: createMethod('DELETE', config),
  }
}
