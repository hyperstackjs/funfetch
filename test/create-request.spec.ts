import { describe, expect, it } from 'vitest'
import { createRequest } from '../src/create-request'

const ep1 = '/api/users'
const functionalEp = ({ company }) => `/api/${company}/listing`
describe('createRequest', () => {
  it('routes', async () => {
    expect(
      createRequest({ ep: ep1, config: {} }, { method: 'GET' })
    ).toMatchSnapshot('simple GET')
    expect(
      createRequest(
        {
          ep: 'https://api.github.com/repos/tannerlinsley/react-query',
          config: {},
        },
        { method: 'GET' }
      )
    ).toMatchSnapshot('absolute URL')
    expect(
      createRequest(
        { ep: ep1, config: {} },
        { method: 'GET', query: { page: 2 } }
      )
    ).toMatchSnapshot('simple GET with query')
    expect(
      createRequest(
        { ep: ep1, config: { base: 'http://microsoft.com' } },
        { method: 'GET', query: { page: 2 } }
      )
    ).toMatchSnapshot('base URL')
    expect(
      createRequest(
        { ep: functionalEp, config: {} },
        { method: 'GET', params: { company: 'microsoft' }, query: { page: 2 } }
      )
    ).toMatchSnapshot('functional endpoint')
    expect(
      createRequest(
        { ep: functionalEp, config: {} },
        {
          method: 'POST',
          data: { user: { name: 'bill' } },
          params: { company: 'microsoft' },
          query: { page: 2 },
        }
      )
    ).toMatchSnapshot('POST data')
    expect(
      createRequest(
        { ep: functionalEp, config: { bearer: 'magic-string' } },
        {
          method: 'POST',
          data: { user: { name: 'bill' } },
          params: { company: 'microsoft' },
          query: { page: 2 },
        }
      )
    ).toMatchSnapshot('use bearer token')
    expect(
      createRequest(
        { ep: functionalEp, config: { bearer: 'magic-string' } },
        {
          method: 'POST',
          data: { user: { name: 'bill' } },
          params: { company: 'microsoft' },
          query: { page: 2 },
          opts: { headers: { Authorization: 'Basic X' } },
        }
      )
    ).toMatchSnapshot('override base opts')
  })
})
