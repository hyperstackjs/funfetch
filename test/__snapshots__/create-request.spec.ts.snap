// Vitest Snapshot v1

exports[`createRequest > routes > POST data 1`] = `
{
  "body": "{\\"user\\":{\\"name\\":\\"bill\\"}}",
  "method": "POST",
  "opts": {
    "headers": {
      "content-type": "application/json",
    },
  },
  "url": "/api/microsoft/listing?page=2",
}
`;

exports[`createRequest > routes > absolute URL 1`] = `
{
  "body": undefined,
  "method": "GET",
  "opts": {
    "headers": {
      "content-type": "application/json",
    },
  },
  "url": "https://api.github.com/repos/tannerlinsley/react-query",
}
`;

exports[`createRequest > routes > base URL 1`] = `
{
  "body": undefined,
  "method": "GET",
  "opts": {
    "headers": {
      "content-type": "application/json",
    },
  },
  "url": "http://microsoft.com/api/users?page=2",
}
`;

exports[`createRequest > routes > functional endpoint 1`] = `
{
  "body": undefined,
  "method": "GET",
  "opts": {
    "headers": {
      "content-type": "application/json",
    },
  },
  "url": "/api/microsoft/listing?page=2",
}
`;

exports[`createRequest > routes > override base opts 1`] = `
{
  "body": "{\\"user\\":{\\"name\\":\\"bill\\"}}",
  "method": "POST",
  "opts": {
    "headers": {
      "Authorization": "Basic X",
      "authorization": "Bearer magic-string",
      "content-type": "application/json",
    },
  },
  "url": "/api/microsoft/listing?page=2",
}
`;

exports[`createRequest > routes > simple GET 1`] = `
{
  "body": undefined,
  "method": "GET",
  "opts": {
    "headers": {
      "content-type": "application/json",
    },
  },
  "url": "/api/users",
}
`;

exports[`createRequest > routes > simple GET with query 1`] = `
{
  "body": undefined,
  "method": "GET",
  "opts": {
    "headers": {
      "content-type": "application/json",
    },
  },
  "url": "/api/users?page=2",
}
`;

exports[`createRequest > routes > use bearer token 1`] = `
{
  "body": "{\\"user\\":{\\"name\\":\\"bill\\"}}",
  "method": "POST",
  "opts": {
    "headers": {
      "authorization": "Bearer magic-string",
      "content-type": "application/json",
    },
  },
  "url": "/api/microsoft/listing?page=2",
}
`;
